src/dotnet/APIView/ClientSPA/src/app/_components/code-panel/code-panel.component.ts

```diff
@@ -619,6 +636,97 @@ export class CodePanelComponent implements OnChanges{
+        let codeLineInfo = convertRowOfTokensToString(row.rowOfTokens);
+        const regex = new RegExp(searchText, "gi");
+        const matches = [...codeLineInfo.matchAll(regex)];
+        if (matches.length > 0) {
+          totalMatches += matches.length;
+          matchedNodeIds.add(row.nodeIdHashed!);
+          this.searchMatchedRowInfo.set(row.nodeIdHashed!, matches);
+        }
+      }
+    });
+    this.highlightSearchMatches();
+    this.codeLineSearchInfoEmitter.emit({ 
+      current: 0, 
+      total: totalMatches,
+      matchedNodeIds: matchedNodeIds
+    });
+  }
+
+  private highlightSearchMatches() {
+    this.clearSearchMatchHighlights();
+    const codeLines = this.elementRef.nativeElement.querySelectorAll('.code-line');
+    
+    codeLines.forEach((codeLine) => {
+      const nodeIdhashed = codeLine.getAttribute('data-node-id');
+      if (nodeIdhashed && this.searchMatchedRowInfo.has(nodeIdhashed)) {
+        const tokens = codeLine.querySelectorAll('.code-line-content > span');
+        const matches = this.searchMatchedRowInfo.get(nodeIdhashed)!;
+  
+        matches.forEach((match) => {
+          const matchStartIndex = match.index!;
+          const matchEndIndex = matchStartIndex + match[0].length;
+  
+          let currentOffset = 0;
+          tokens.forEach((token) => {
+            const tokenText = token.textContent || '';
+            const tokenLength = tokenText.length;
+
+            const tokenStart = currentOffset;
+            const tokenEnd = currentOffset + tokenLength;
+  
+            if (matchStartIndex < tokenEnd && matchEndIndex > tokenStart) {
+              const highlightStart = Math.max(0, matchStartIndex - tokenStart);
+              const highlightEnd = Math.min(tokenLength, matchEndIndex - tokenStart);
+
+              const beforeMatch = tokenText.slice(0, highlightStart);
+              const matchText = tokenText.slice(highlightStart, highlightEnd);
+              const afterMatch = tokenText.slice(highlightEnd);
+
+              token.innerHTML = `${beforeMatch}<mark class="codeline-search-match-highlight">${matchText}</mark>${afterMatch}`;
+            }
+  
+            currentOffset += tokenLength;
+          });
+        });
+      }
+    });
+  }
+
+  /**
+   * Navigates to the next or previous code line that contains a search match but is outside the viewport
+   */
+  private navigateToCodeLineWithSearchMatch(previousPosition: number, newPosition: number) {
+    if (newPosition > previousPosition) {
+      // Find the next code line that contains a search match but is below the visible viewport
+      return;
+    } else if (newPosition < previousPosition) {
+      // Find the previous code line that contains a search match but is above the visible viewport
+      return;
+    }
+  }
+
+  private clearSearchMatchHighlights() {
+    this.elementRef.nativeElement.querySelectorAll('.codeline-search-match-highlight').forEach((element) => {
```

Comment:
should it clear the map `searchMatchedRowInfo` also?

Reply:
Not in all cases. During scrolling we want to only clear the highlight then highlight.